File: Chapter 04/Java/src/main/java/com/grokkingsoftwarearchitecture/chapter04/section_4_2_downward_dependency/after/OrderService.java
package com.grokkingsoftwarearchitecture.chapter04.section_4_2_downward_dependency.after;

/**
 * BUSINESS LOGIC LAYER.
 * ARCHITECTURE NOTE: This service is "ignorant" of the database.
 * It only knows about the OrderRepository interface.
 */
public class OrderService {
    private final OrderRepository repo; 

    public OrderService(OrderRepository repo) { 
        this.repo = repo; 
    }

    public void saveOrder(Order order) {
        // Calls DOWNWARDS via interface 
        repo.save(order);
    }
}

File: Chapter 04/Python/section_4_2_downward_dependency/after/order_service.py
from .order_repository import OrderRepository
from .order import Order

class OrderService:
    """
    BUSINESS LOGIC LAYER (HIGH-LEVEL POLICY).
    ARCHITECTURE NOTE: This service is 'UI-Ignorant' and 
    'DB-Ignorant'. It only knows about the Repository 
    abstraction. This follows the Downward Dependency rule.
    """
    def __init__(self, repo: OrderRepository):
        # Dependency Injection: We receive the tool we need.
        self.repo = repo 

    def save_order(self, order: Order):
        # We call DOWNWARD into the abstraction.
        self.repo.save(order)